File: shared-core/src/commonMain/kotlin/com/appversal/appstorys/core/model/SpinWheelModels.kt
package com.appversal.appstorys.core.model

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@Serializable
data class SpinTheWheelDetails(
    val id: String? = null,
    // Direct fields from backend
    val popupTitle: String? = null,
    val popupDescription: String? = null,
    val spinButtonText: String? = null,
    val availableSpins: Int? = null,
    val enableBackdrop: Boolean? = null,
    val rewardDisplayMode: String? = null, // "pop-up" or other modes
    val slices: List<WheelSlice>? = null,
    val content: SpinWheelContent? = null,
    val styling: SpinWheelStyling? = null,
    val link: String? = null, // Redirect link for CTA button
    val button_text: String? = null // Alternative button text field
) : CampaignDetails()

@Serializable
data class SpinWheelContent(
    val availableSpinsText: String? = null,
    val rewardConfiguration: SpinWheelRewardConfig? = null,
    val userInteraction: SpinWheelUserInteraction? = null,
    val wheelConfiguration: SpinWheelWheelConfig? = null
)

@Serializable
data class SpinWheelRewardConfig(
    val rewardEnableBackdrop: Boolean? = null,
    val rewardPopupDescription: String? = null,
    val rewardPopupTitle: String? = null
)

@Serializable
data class SpinWheelUserInteraction(
    val hapticFeedback: Boolean? = null,
    val numberSpin: Int? = null
)

@Serializable
data class SpinWheelWheelConfig(
    val spinDirection: String? = null, // "clockwise" or "anti-clockwise"
    val totalSliceCount: Int? = null
)

@Serializable
data class WheelSlice(
    val id: String? = null,
    val parent: String? = null,
    val coupon: String? = null,
    val noPrize: Boolean? = null,
    val prizeLabel: String? = null,
    @SerialName("subText") val subText: String? = null, // Optional supporting text shown below the prize name
    val sliceMedia: String? = null,
    val weight: Int? = null,
    val styling: WheelSliceStyling? = null,
    val link: String? = null, // Redirect link for this specific prize (Button Redirect To)
    @SerialName("buttonCtaText") val buttonCtaText: String? = null, // Custom text for primary CTA button
    @SerialName("tncCtaText") val tncCtaText: String? = null, // Custom text for Terms & Conditions link
    @SerialName("termsAndConditions") val termsAndConditions: String? = null, // Content displayed when user clicks "Know More to Claim"
    val rewards: List<SliceReward>? = null // Array of reward configurations for this slice
)

// Slice Reward Configuration (per-slice reward display settings)
@Serializable
data class SliceReward(
    val id: String? = null,
    val sliceId: String? = null,
    val prizeName: String? = null,
    val couponCode: String? = null,
    val subText: String? = null,
    val buttonCta: String? = null,
    @SerialName("tNcCta") val tNcCta: String? = null,
    @SerialName("termsNConditions") val termsNConditions: String? = null,
    val link: String? = null,
    val sliceRewardMedia: String? = null,
    val styling: SliceRewardStyling? = null
)

@Serializable
data class SliceRewardStyling(
    val couponCodeCta: SliceRewardCtaStyling? = null,
    val cta: SliceRewardCtaStyling? = null,
    val priceLabel: SliceRewardTextConfig? = null,
    val subtitleText: SliceRewardTextConfig? = null
)

@Serializable
data class SliceRewardCtaStyling(
    val container: SliceRewardCtaContainer? = null,
    val cornerRadius: CornerRadius? = null,
    val margin: WheelMargin? = null,
    val text: SliceRewardCtaText? = null
)

@Serializable
data class SliceRewardCtaContainer(
    val alignment: String? = null,
    val backgroundColor: String? = null,
    val borderColor: String? = null,
    val borderWidth: Int? = null,
    val ctaFullWidth: Boolean? = null,
    val ctaWidth: Int? = null,
    val height: Int? = null
)

@Serializable
data class SliceRewardCtaText(
    val color: String? = null,
    val fontDecoration: List<String>? = null,
    val fontFamily: String? = null,
    @Serializable(with = NullableIntSerializer::class)
    val fontSize: Int? = null
)

@Serializable
data class SliceRewardTextConfig(
    val textStyle: SliceRewardTextStyle? = null
)

@Serializable
data class SliceRewardTextStyle(
    val color: String? = null,
    val fontDecoration: List<String>? = null,
    val fontFamily: String? = null,
    @Serializable(with = NullableIntSerializer::class)
    val fontSize: Int? = null,
    val margin: WheelMargin? = null,
    val textAlign: String? = null
)

// Styling Models (from Dashboard Styling Tab)
@Serializable
data class SpinWheelStyling(
    val spinTheWheel: SpinTheWheelMainStyling? = null,
    val rewardConfiguration: WheelRewardStyling? = null
)

@Serializable
data class SpinTheWheelMainStyling(
    val crossButton: WheelCrossButtonConfig? = null,
    val visualTextCommunication: WheelVisualTextStyling? = null,
    val wheelConfiguration: WheelConfigurationStyling? = null
)

@Serializable
data class WheelVisualTextStyling(
    val title: WheelTextStyleConfig? = null,
    val subtitle: WheelTextStyleConfig? = null,
    val availableSpinText: WheelAvailableSpinTextStyle? = null,
    val spinButton: WheelSpinButtonStyle? = null,
    val backdropColor: String? = null,
    val backdropOpacity: Int? = null
)

@Serializable
data class WheelAvailableSpinTextStyle(
    val textStyle: WheelTextStyleDetails? = null
)

@Serializable
data class WheelSpinButtonStyle(
    val container: WheelButtonContainerStyle? = null,
    val text: WheelButtonTextStyle? = null,
    val margin: WheelMargin? = null
)

@Serializable
data class WheelButtonContainerStyle(
    val alignment: String? = null,
    val backgroundColor: String? = null,
    val borderColor: String? = null,
    val borderWidth: Int? = null,
    val cornerRadius: CornerRadius? = null,
    val fullWidth: Boolean? = null,
    val height: Int? = null,
    val width: Int? = null
)

@Serializable
data class WheelButtonTextStyle(
    val color: String? = null,
    val fontFamily: String? = null,
    val fontSize: Int? = null,
    val fontWeight: String? = null,
    val fontStyle: String? = null,
    val fontDecoration: List<String>? = null
)

@Serializable
data class WheelConfigurationStyling(
    val backgroundColor: String? = null,
    val borderColor: String? = null,
    val borderWidth: Int? = null,
    val backgroundImage: String? = null,
    val backgroundImageOpacity: Float? = null,
    val size: Int? = null // Wheel diameter in dp - defaults to 300 if not set
)

@Serializable
data class WheelRewardStyling(
    val backdropColor: String? = null,
    val cardBackgroundColor: String? = null,
    val confetti: WheelConfettiConfig? = null,
    val crossButton: WheelCrossButtonConfig? = null,
    val title: WheelTextStyleConfig? = null,
    val subtitle: WheelTextStyleConfig? = null
)

@Serializable
data class WheelConfettiConfig(
    val selectedStyle: String? = null,
    val color: WheelButtonColors? = null,
    val image: String? = null,
    val margin: WheelMargin? = null,
    val padding: WheelMargin? = null
)

@Serializable
data class WheelCrossButtonConfig(
    val enabled: Boolean? = null,
    val selectedStyle: String? = null,
    val size: Int? = null,
    val color: WheelButtonColors? = null,
    val image: String? = null,
    val margin: WheelMargin? = null,
    val alignment: String? = null
)

@Serializable
data class WheelButtonColors(
    val cross: String? = null,
    val fill: String? = null,
    val stroke: String? = null
)

@Serializable
data class WheelMargin(
    val top: Int? = null,
    val bottom: Int? = null,
    val left: Int? = null,
    val right: Int? = null
)

@Serializable
data class WheelTextStyleConfig(
    val textStyle: WheelTextStyleDetails? = null
)

@Serializable
data class WheelTextStyleDetails(
    val color: String? = null,
    val fontFamily: String? = null,
    val fontSize: Int? = null,
    val fontWeight: String? = null,
    val fontStyle: String? = null,
    val fontDecoration: List<String>? = null,
    val textAlign: String? = null,
    val margin: WheelMargin? = null
)

@Serializable
data class WheelSliceStyling(
    val wheelStyling: WheelSliceWheelStyling? = null
)

@Serializable
data class WheelSliceWheelStyling(
    val color: WheelSliceColors? = null,
    val cornerRadius: CornerRadius? = null,
    val strokeWidth: Int? = null,
    val margin: WheelMargin? = null,
    val image: WheelSliceImageStyling? = null,
    val priceLabel: WheelSlicePriceLabelStyling? = null
)

@Serializable
data class WheelSliceColors(
    val background: String? = null,
    val stroke: String? = null
)

@Serializable
data class WheelSliceImageStyling(
    val cornerRadius: CornerRadius? = null,
    val rotation: Int? = null
)

@Serializable
data class WheelSlicePriceLabelStyling(
    val margin: WheelMargin? = null,
    val textStyle: WheelSliceTextStyle? = null
)

@Serializable
data class WheelSliceTextStyle(
    val font: String? = null,
    val fontSize: Int? = null,
    val color: String? = null
)

File: appstorys_flutter/lib/src/models/spin_wheel_models.dart
// Models for Spin The Wheel (STW) campaign type.

// ── Helpers ───────────────────────────────────────────────────────────────────

double _toDouble(dynamic v, [double def = 0.0]) {
  if (v == null) return def;
  if (v is num) return v.toDouble();
  if (v is String) return double.tryParse(v) ?? def;
  return def;
}

int _toInt(dynamic v, [int def = 0]) {
  if (v == null) return def;
  if (v is int) return v;
  if (v is num) return v.toInt();
  if (v is String) return int.tryParse(v) ?? def;
  return def;
}

bool _toBool(dynamic v, [bool def = false]) {
  if (v == null) return def;
  if (v is bool) return v;
  return def;
}

String _toStr(dynamic v, [String def = '']) {
  if (v == null) return def;
  return v.toString();
}

Map<String, dynamic> _safeMap(dynamic v) =>
    v is Map ? Map<String, dynamic>.from(v) : <String, dynamic>{};

// ── Top-level Campaign ────────────────────────────────────────────────────────

class SpinWheelCampaign {
  final String id;
  final String campaignType;
  final SpinWheelDetails details;

  SpinWheelCampaign({
    required this.id,
    required this.campaignType,
    required this.details,
  });

  factory SpinWheelCampaign.fromJson(Map<String, dynamic> json) {
    final id = (json['campaign_id'] ?? json['id'] ?? '').toString();
    return SpinWheelCampaign(
      id: id,
      campaignType: _toStr(json['campaign_type'], 'STW'),
      details: SpinWheelDetails.fromJson(
        _safeMap(json['details']),
        fallbackId: id,
      ),
    );
  }
}

// ── Campaign Details ──────────────────────────────────────────────────────────

class SpinWheelDetails {
  final String id;

  // Content fields
  final int availableSpins;
  final String popupTitle;
  final String popupDescription;
  final String spinButtonText;
  final String rewardDisplayMode;
  final String availableSpinsText;
  final String spinDirection;
  final int totalSliceCount;
  final bool hapticFeedback;

  // Backdrop
  final String backdropColor;
  final int backdropOpacity;

  // Wheel-screen text/button styling (raw textStyle maps from VTC)
  final Map<String, dynamic> titleTextStyle;
  final Map<String, dynamic> subtitleTextStyle;
  final Map<String, dynamic> spinsTextStyle;

  // Spin button (full spinButton map: container, margin, text)
  final Map<String, dynamic> spinButton;

  // Wheel border styling
  final Map<String, dynamic> wheelConfig;

  // Cross button on wheel screen
  final Map<String, dynamic> stwCrossButton;

  // Reward screen
  final String rewardBackdropColor;
  final Map<String, dynamic> rewardTitleStyle;
  final Map<String, dynamic> rewardSubtitleStyle;
  final Map<String, dynamic> rewardCrossButton;
  final Map<String, dynamic> confetti;

  final List<SpinWheelSlice> slices;

  SpinWheelDetails({
    required this.id,
    required this.availableSpins,
    required this.popupTitle,
    required this.popupDescription,
    required this.spinButtonText,
    required this.rewardDisplayMode,
    required this.availableSpinsText,
    required this.spinDirection,
    required this.totalSliceCount,
    required this.hapticFeedback,
    required this.backdropColor,
    required this.backdropOpacity,
    required this.titleTextStyle,
    required this.subtitleTextStyle,
    required this.spinsTextStyle,
    required this.spinButton,
    required this.wheelConfig,
    required this.stwCrossButton,
    required this.rewardBackdropColor,
    required this.rewardTitleStyle,
    required this.rewardSubtitleStyle,
    required this.rewardCrossButton,
    required this.confetti,
    required this.slices,
  });

  factory SpinWheelDetails.fromJson(
    Map<String, dynamic> json, {
    String fallbackId = '',
  }) {
    final content = _safeMap(json['content']);
    final wheelCfgContent = _safeMap(content['wheelConfiguration']);
    final userInt = _safeMap(content['userInteraction']);

    final styling = _safeMap(json['styling']);
    final stw = _safeMap(styling['spinTheWheel']);
    final vtc = _safeMap(stw['visualTextCommunication']);

    final titleMap = _safeMap(vtc['title']);
    final titleTs = _safeMap(titleMap['textStyle']);

    final subtitleMap = _safeMap(vtc['subtitle']);
    final subtitleTs = _safeMap(subtitleMap['textStyle']);

    final spinsTextMap = _safeMap(vtc['availableSpinText']);
    final spinsTextTs = _safeMap(spinsTextMap['textStyle']);

    final spinBtn = _safeMap(vtc['spinButton']);
    final wheelCfgStl = _safeMap(stw['wheelConfiguration']);
    final stwCross = _safeMap(stw['crossButton']);

    final rewardCfg = _safeMap(styling['rewardConfiguration']);
    final rwTitle = _safeMap(rewardCfg['title']);
    final rwTitleTs = _safeMap(rwTitle['textStyle']);
    final rwSubtitle = _safeMap(rewardCfg['subtitle']);
    final rwSubtitleTs = _safeMap(rwSubtitle['textStyle']);
    final rwCross = _safeMap(rewardCfg['crossButton']);
    final confettiMap = _safeMap(rewardCfg['confetti']);

    final rawSlices = json['slices'];
    final slices = rawSlices is List
        ? rawSlices
            .whereType<Map>()
            .map((s) => SpinWheelSlice.fromJson(Map<String, dynamic>.from(s)))
            .toList()
        : <SpinWheelSlice>[];

    return SpinWheelDetails(
      id: _toStr(json['id'], fallbackId),
      availableSpins: _toInt(json['availableSpins'], 1),
      popupTitle: _toStr(json['popupTitle'], 'Spin To Win'),
      popupDescription: _toStr(json['popupDescription']),
      spinButtonText: _toStr(json['spinButtonText'], 'Spin now'),
      rewardDisplayMode: _toStr(json['rewardDisplayMode'], 'full-screen'),
      availableSpinsText: _toStr(content['availableSpinsText'], 'Spins left'),
      spinDirection: _toStr(wheelCfgContent['spinDirection'], 'clockwise'),
      totalSliceCount: _toInt(wheelCfgContent['totalSliceCount'], slices.length),
      hapticFeedback: _toBool(userInt['hapticFeedback']),
      backdropColor: _toStr(vtc['backdropColor'], '#000000'),
      backdropOpacity: _toInt(vtc['backdropOpacity'], 50),
      titleTextStyle: titleTs,
      subtitleTextStyle: subtitleTs,
      spinsTextStyle: spinsTextTs,
      spinButton: spinBtn,
      wheelConfig: wheelCfgStl,
      stwCrossButton: stwCross,
      rewardBackdropColor: _toStr(rewardCfg['backdropColor'], '#000000ff'),
      rewardTitleStyle: rwTitleTs,
      rewardSubtitleStyle: rwSubtitleTs,
      rewardCrossButton: rwCross,
      confetti: confettiMap,
      slices: slices,
    );
  }
}

// ── Wheel Slice ───────────────────────────────────────────────────────────────

class SpinWheelSlice {
  final String id;
  final String prizeLabel;
  final String sliceMedia;
  final bool noPrize;
  final int weight;

  // Wheel visual styling
  final String bgColor;
  final String strokeColor;
  final double strokeWidth;
  final double cornerRadiusTopLeft;
  final double cornerRadiusTopRight;
  final double cornerRadiusBottomLeft;
  final double cornerRadiusBottomRight;

  // Prize label text
  final String labelColor;
  final String labelFontFamily;
  final double labelFontSize;

  // Image config
  final double imageRotation;

  final SpinWheelReward? reward;

  SpinWheelSlice({
    required this.id,
    required this.prizeLabel,
    required this.sliceMedia,
    required this.noPrize,
    required this.weight,
    required this.bgColor,
    required this.strokeColor,
    required this.strokeWidth,
    required this.cornerRadiusTopLeft,
    required this.cornerRadiusTopRight,
    required this.cornerRadiusBottomLeft,
    required this.cornerRadiusBottomRight,
    required this.labelColor,
    required this.labelFontFamily,
    required this.labelFontSize,
    required this.imageRotation,
    this.reward,
  });

  factory SpinWheelSlice.fromJson(Map<String, dynamic> json) {
    final styling = _safeMap(json['styling']);
    final ws = _safeMap(styling['wheelStyling']);
    final color = _safeMap(ws['color']);
    final corner = _safeMap(ws['cornerRadius']);
    final plLabel = _safeMap(ws['priceLabel']);
    final plTs = _safeMap(plLabel['textStyle']);
    final image = _safeMap(ws['image']);

    final rewards = json['rewards'];
    final reward = rewards is List && rewards.isNotEmpty
        ? SpinWheelReward.fromJson(
            Map<String, dynamic>.from(rewards.first as Map))
        : null;

    return SpinWheelSlice(
      id: _toStr(json['id']),
      prizeLabel: _toStr(json['prizeLabel']),
      sliceMedia: _toStr(json['sliceMedia']),
      noPrize: _toBool(json['noPrize']),
      weight: _toInt(json['weight'], 10),
      bgColor: _toStr(color['background'], '#0B1020'),
      strokeColor: _toStr(color['stroke'], '#FFFFFF'),
      strokeWidth: _toDouble(ws['strokeWidth'], 2),
      cornerRadiusTopLeft: _toDouble(corner['topLeft']),
      cornerRadiusTopRight: _toDouble(corner['topRight']),
      cornerRadiusBottomLeft: _toDouble(corner['bottomLeft']),
      cornerRadiusBottomRight: _toDouble(corner['bottomRight']),
      labelColor: _toStr(plTs['color'], '#FFFFFF'),
      labelFontFamily: _toStr(plTs['fontFamily'], 'Helvetica'),
      labelFontSize: _toDouble(plTs['fontSize'], 12),
      imageRotation: _toDouble(image['rotation']),
      reward: reward,
    );
  }
}

// ── Slice Reward ──────────────────────────────────────────────────────────────

class SpinWheelReward {
  final String id;
  final String sliceId;
  final String prizeName;
  final String couponCode;
  final String buttonCta;
  final String subText;
  final String sliceRewardMedia;
  final String link;
  final String termsNConditions;
  final String tNcCta;

  // Raw styling maps — parsed inline in the widget
  final Map<String, dynamic> ctaStyling;
  final Map<String, dynamic> couponStyling;
  final Map<String, dynamic> priceLabelTextStyle;
  final Map<String, dynamic> subtitleTextStyle;

  SpinWheelReward({
    required this.id,
    required this.sliceId,
    required this.prizeName,
    required this.couponCode,
    required this.buttonCta,
    required this.subText,
    required this.sliceRewardMedia,
    required this.link,
    required this.termsNConditions,
    required this.tNcCta,
    required this.ctaStyling,
    required this.couponStyling,
    required this.priceLabelTextStyle,
    required this.subtitleTextStyle,
  });

  factory SpinWheelReward.fromJson(Map<String, dynamic> json) {
    final styling = _safeMap(json['styling']);
    final cta = _safeMap(styling['cta']);
    final coupon = _safeMap(styling['couponCodeCta']);
    final priceLabel = _safeMap(styling['priceLabel']);
    final priceLabelTs = _safeMap(priceLabel['textStyle']);
    final subtitleMap = _safeMap(styling['subtitleText']);
    final subtitleTs = _safeMap(subtitleMap['textStyle']);

    return SpinWheelReward(
      id: _toStr(json['id']),
      sliceId: _toStr(json['sliceId']),
      prizeName: _toStr(json['prizeName']),
      couponCode: _toStr(json['couponCode']),
      buttonCta: _toStr(json['buttonCta'], 'Claim'),
      subText: _toStr(json['subText']),
      sliceRewardMedia: _toStr(json['sliceRewardMedia']),
      link: _toStr(json['link']),
      termsNConditions: _toStr(json['termsNConditions']),
      tNcCta: _toStr(json['tNcCta'], 'T&C'),
      ctaStyling: cta,
      couponStyling: coupon,
      priceLabelTextStyle: priceLabelTs,
      subtitleTextStyle: subtitleTs,
    );
  }
}
